### PR TITLE
Include cursor in textfield intrinsic width measurement

### LINKED ISSUE #24612: TextField's IntrinsicWidth should take cursor into account
When computing the intrinsic size, I think, the cursor width should be taken into account so that the text field doesn't scroll if the cursor is at the end. I made the cursor very wide to illustrate the point. The text field below contains `12345` and is sized using an `InstrinsicWidth` widget. The `1` has been scrolled to the left to make room for the cursor. The same effect occurs with a "normal" cursor, but is barely visible.

image

If you consider this behavior "by design", a way to add a width and/or height offset to the intrinsic widgets would be a workaround to get rid of the scrolling.

## Output of flutter doctor

```
[✓] Flutter (Channel master, v0.11.10-pre.1, on Mac OS X 10.14.1 18B75, locale de-DE)
[✓] Android toolchain - develop for Android devices (Android SDK 28.0.3)
[!] iOS toolchain - develop for iOS devices (Xcode 10.1)
    ✗ Verify that all connected devices have been paired with this computer in Xcode.
      If all devices have been paired, libimobiledevice and ideviceinstaller may require updating.
      To update with Brew, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew uninstall --ignore-dependencies usbmuxd
        brew install --HEAD usbmuxd
        brew unlink usbmuxd
        brew link usbmuxd
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
[✓] Android Studio (version 3.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
[✓] VS Code (version 1.29.1)
[✓] Connected device (2 available)
```

### PR DESCRIPTION
**Pull request by justinmc**
Closes https://github.com/flutter/flutter/issues/24612

When using `TextField` with `InstrinsicWidth`, the width of the cursor was not included in the measurement of the width of the text content.  This caused content to scroll slightly instead of triggering a bump up to the next width step size.  Here the string reads "123456", but it has been scrolled so that the 1 is not visible:

</img>

This PR includes the cursor width in the calculation, so the width will increase before scrolling happens.  Here I've used a large step size to illustrate the bump up in size with the same string:

</img>

### CODE REVIEW

**Comment by justinmc** [comment]
@HansMuller What do you think of this solution, any implications that I'm not thinking of?  There was only 1 test that failed when I made this change (see the 1 pixel that I changed in a test).

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -1006,7 +1006,7 @@ class RenderEditable extends RenderBox {
   @override
   double computeMaxIntrinsicWidth(double height) {
     _layoutText(double.infinity);
-    return _textPainter.maxIntrinsicWidth;
+    return _textPainter.maxIntrinsicWidth + cursorWidth;
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/editable.dart]: We should probably `assert(cursorWidth != null && cursorWidth >= 0.0)` in the RenderEditable constructor. There are many other places in the code that appear to make this assumption.  Not sure if cursorWidth == 0.0 really works; if not assert that it's > 0.0.

File: packages/flutter/test/material/text_field_test.dart [hunk 2]
```diff
@@ -3382,4 +3382,44 @@ void main() {
     await tester.tap(find.byType(TextField));
     expect(tapCount, 0);
   });
+
+  testWidgets('Includes cursor for TextField', (WidgetTester tester) async {
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: We should also verify that if the the TextField is created within an IntrinsicWidth with no stepWidth specified, that the result is a textfield that's just wide enough to drive to the cursor to the end without scrolling. 

Tests use the fixed-width Ahem font by default, so you can easily predict the textfield's exact width.

Should check this with a textfield that's `textAlign: TextAlign.right` and `TextAlign.left`

File: packages/flutter/test/rendering/editable_test.dart [hunk 3]
```diff
@@ -38,7 +38,7 @@ void main() {
       textSelectionDelegate: delegate,
     );
     expect(editable.getMinIntrinsicWidth(double.infinity), 50.0);
-    expect(editable.getMaxIntrinsicWidth(double.infinity), 50.0);
+    expect(editable.getMaxIntrinsicWidth(double.infinity), 51.0);
```
Comment (HansMuller) [inline packages/flutter/test/rendering/editable_test.dart]: Should add a comment here about including the the 1.0 cursorWidth

**Review by HansMuller — COMMENTED** [review]
This is looking good. Just needs a little more test coverage

File: packages/flutter/lib/src/rendering/editable.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/rendering/editable.dart]: It looks like cursorWidth 0.0 works and results in not showing the cursor at all, so I'm using `>=` in the assertion.

**Comment by justinmc** [comment]
@HansMuller Done and ready for re-review

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
This had some unfortunate impact on some of the performance metrics.

**Comment by justinmc** [comment]
Hmm, I'll try to dig into that.

**Comment by justinmc** [comment]
For the record this code change seems to be absolved of causing performance problems now.  Performance is back to normal after something was changed with the bots.

**Pull request merged**